policy/policy.go

```diff
@@ -681,3 +689,21 @@ func (s *GroupType) UnmarshalJSON(data []byte) error {
 
 	return errors.New("failed to unmarshal group type: " + str)
 }
+
+func variantsExecutionChecksum(q *explorer.Mquery, c checksums.Fast, includeImpact bool, getQuery func(ctx context.Context, mrn string) (*explorer.Mquery, error)) (checksums.Fast, error) {
+	// TODO: this function needs to be safe against cycles
```

Comment:
Do you mean that the same `CodeId` shouldn't be added to the checksum multiple times? If yes, then I guess this change should be sufficient:
```go
func variantsExecutionChecksum(q *explorer.Mquery, c checksums.Fast, includeImpact bool, getQuery func(ctx context.Context, mrn string) (*explorer.Mquery, error), added map[string]struct{}) (checksums.Fast, error) {
	if _, ok := added[q.CodeId]; ok {
		return c, nil
	}

	c = c.Add(q.CodeId)
	visited[q.CodeId] = struct{}{}
	if includeImpact {
		c = c.AddUint(q.Impact.Checksum())
	}

	for _, ref := range q.Variants {
		if v, err := getQuery(context.Background(), ref.Mrn); err == nil {
			return variantsExecutionChecksum(v, c, includeImpact, getQuery, added)
		} else {
			return 0, errors.New("cannot find dependent composed query '" + ref.Mrn + "'")
		}
	}
	return c, nil
}
```

Reply:
its not really about being added so much as we dont want cycles because then the function wouldn't terminate

Reply:
also, its possible that there is not a code id because there is no code